### PR TITLE
Show errors on why the registration of a user fails

### PR DESCRIPTION
**Pull request by Steffengreiner**
Addresses DM-361 by adding a full name policy to check for user input in the full name field. 
Additionally introduces distinct error messaging and component handling dependent on the violated domain policy.

### CODE REVIEW

File: domain/src/main/java/life/qbic/domain/usermanagement/User.java [hunk 1]
```diff
@@ -55,7 +55,8 @@ protected User() {}
    */
   public static User create(String fullName, String email) throws UserException {
     String uuid = String.valueOf(UUID.randomUUID());
-    var user = new User(fullName);
+    var user = new User();
+    user.setFullName(fullName);
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/domain/usermanagement/User.java:59]: Why did you change that?

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * <b>Full Name Policy</b>
+ *
+ * <p>Validates if a String is concurrent with the full name requirements
+ *
+ * @since 1.0.0
+ */
+public class FullNamePolicy {
+
+  private static final int MIN_LENGTH = 1;
+
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java:13]: Why did you add that? AFAIK we dont have a policy there other than that it needs to be provided

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java [hunk 3]
```diff
@@ -57,30 +60,16 @@ private void addListener() {
...
-    userRegistrationLayout.passwordTooShortMessage.setVisible(false);
     userRegistrationLayout.alreadyUsedEmailMessage.setVisible(false);
+    userRegistrationLayout.errorMessage.setVisible(false);
+    userRegistrationLayout.passwordTooShortMessage.setVisible(false);
+  }
+
+  private void resetComponentErrors() {
+    userRegistrationLayout.fullName.setErrorMessage("A username must be at least 1 character");
+    userRegistrationLayout.email.setErrorMessage("Please provide a valid email");
+    userRegistrationLayout.password.setErrorMessage("Password too short");
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:72]: I am confused, this reads as if you set the error rather than reset

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java [hunk 4]
```diff
@@ -90,9 +79,30 @@ public void onSuccess() {
...
+    }
+    if (reason.equals("User with email address already exists.")) {
+      userRegistrationLayout.alreadyUsedEmailMessage.setVisible(true);
+      userRegistrationLayout.email.setInvalid(true);
+      userRegistrationLayout.email.setErrorMessage(reason);
+    }
+    if (reason.equals("Unexpected error occurred.")) {
+      userRegistrationLayout.errorMessage.setVisible(true);
+    }
+    if (reason.equals("Password shorter than 8 characters.")) {
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:102]: Screams for a switch statement

**Review by jenniferboedker — CHANGES_REQUESTED** [review]
Thanks for your contribution! I am a little surprised about the new policy that you have added. Maybe you can elaborate on how you did that. I am not sure if that is a policy in our case since we require the value only to be present. But I am happy to discuss that with you :) 
Other than that: did something change with the application properties? I cannot run the application on my laptop to test if it works as desired 😿

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java (hunk 3 above)
Comment (Steffengreiner) [reply webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java]: Good Point i renamed the method to make it's purpose more clear 👍

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/EmailFormatPolicy.java [hunk 5]
```diff
@@ -36,7 +38,7 @@ public static EmailFormatPolicy create() {
    */
   public PolicyCheckReport validate(String email) {
     if (!honoursRFCSpec(email)) {
-      return new PolicyCheckReport(PolicyStatus.FAILED, "Invalid email address format.");
+      return new PolicyCheckReport(PolicyStatus.FAILED, InvalidEmailMessage);
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/EmailFormatPolicy.java:41]: Originally i wanted to have a link between the invalid email messages the policy could trigger and the ones recognized by the switch statement in the handler to avoid duplication here.
 But I'm unsure how to do this without violating the clean architecture principles.

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java (hunk 2 above)
Comment (Steffengreiner) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java:13]: I think a policy is necessary here since the Domain logic should decide on the validity of the input. 
Also this allows us to test our requirements easily and adapt the policy if necessary down the line. 
We could e.g. think about not allowing nonsensical names down the line as suggested e.g. here: https://support.google.com/a/answer/9193374?hl=en

File: domain/src/main/java/life/qbic/domain/usermanagement/User.java (hunk 1 above)
Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/domain/usermanagement/User.java:59]: I would prefer to keep it as it was since it is no real policy to check for the presence of a value.

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/EmailFormatPolicy.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/EmailFormatPolicy.java:41]: ok but then you can change it back, right?

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java (hunk 2 above)
Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java:13]: As I said above I dont see it as a policy. Its a check for presence, the other things are requirements you just came up with. Its nothing that we have for now. Therfore I would kindly ask you to remove the policy and check the value in the user entity (I think this was done like this before, right?). You can also do it in the use case but a policy is overkill for this

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/PasswordPolicy.java [hunk 6]
```diff
@@ -32,7 +34,7 @@ public static PasswordPolicy create() {
    */
   public PolicyCheckReport validate(char[] rawPassword) {
     if (rawPassword.length < MIN_LENGTH) {
-      return new PolicyCheckReport(PolicyStatus.FAILED, "Password shorter than 8 characters.");
+      return new PolicyCheckReport(PolicyStatus.FAILED, invalidPasswordMessage);
```
Comment (jenniferboedker) [inline domain/src/main/java/life/qbic/domain/usermanagement/policies/PasswordPolicy.java:37]: same here as above. no need for a variable

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java [hunk 7]
```diff
@@ -90,9 +79,30 @@ public void onSuccess() {
...
+        userRegistrationLayout.alreadyUsedEmailMessage.setVisible(true);
+        userRegistrationLayout.email.setInvalid(true);
+        userRegistrationLayout.email.setErrorMessage(reason);
+      }
+      case "Unexpected error occurred." -> userRegistrationLayout.errorMessage.setVisible(true);
+      case "Password shorter than 8 characters." -> {
+        userRegistrationLayout.passwordTooShortMessage.setVisible(true);
+        userRegistrationLayout.password.setInvalid(true);
+        userRegistrationLayout.password.setErrorMessage(reason);
+      }
```
Comment (jenniferboedker) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:105]: default case is missing. we could show an unexpected error or so :)

Comment (KochTobi) [reply webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:105]: Short remark, feel free to ignore:
Here an enum of possible reasons could make sense, you would not need a default then.

Comment (Steffengreiner) [reply webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:90]: In my dreamworld i'd define the reason set in the domain logic in an enum which can also be used in the handler for checking. But this would violate clean architecture 🤷‍♂️

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java (hunk 2 above)
Comment (Steffengreiner) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/FullNamePolicy.java:13]: I don't see the problem with having an easily extendable policy based solution that checks the user input. 
IMO the policy should contain ALL requirements for the associated user field.
Before the introduction of the policy we did not check the user input at all in the user entity.

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/PasswordPolicy.java (hunk 6 above)
Comment (Steffengreiner) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/PasswordPolicy.java:37]: alright 👍

File: domain/src/main/java/life/qbic/domain/usermanagement/policies/EmailFormatPolicy.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply domain/src/main/java/life/qbic/domain/usermanagement/policies/EmailFormatPolicy.java:41]: Sure 👍

File: domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java [hunk 8]
```diff
@@ -34,15 +36,51 @@ public UserDomainService(UserRepository userRepository) {
    */
   public void createUser(String fullName, String email, char[] rawPassword) throws UserException {
     // First check, if a user with the provided email already exists
-    if (userRepository.findByEmail(email).isPresent()) {
-      throw new UserException("User with email address already exists.");
-    }
+    validateUserCredentials(fullName, email, rawPassword);
     var domainEventPublisher = DomainEventPublisher.instance();
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java:40]: What you do here is a different behaviour then to lookup whether a user already exists. You could replace that with `isEmailInDatabase` but I do not think that it adds readability here.

File: domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java [hunk 9]
```diff
@@ -34,15 +36,51 @@ public UserDomainService(UserRepository userRepository) {
...
+    } catch (UserException e) {
+      userExceptions.add(e);
+    }
+    try {
+      user.setPassword(rawPassword);
+    } catch (UserException e) {
+      userExceptions.add(e);
+    }
+    throwCorrectUserException(userExceptions);
+  }
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java:74]: This does not belong here. The user itself has the responsibility to check it's parameters upon creation. That is why we have a static method in the user for creating users. We do not intent the constructor of a User to be used outside of the User class.

File: domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java [hunk 10]
```diff
@@ -34,15 +36,51 @@ public UserDomainService(UserRepository userRepository) {
...
+  private void throwCorrectUserException(List<UserException> userExceptions) {
+    if (userExceptions.size() == 1) {
+      throw userExceptions.get(0);
+    }
+    if (userExceptions.size() > 1) {
+      UserException groupedUserException = new UserException("Multiple Invalid Credentials");
+      userExceptions.forEach(groupedUserException::addSuppressed);
+      throw groupedUserException;
+    }
+  }
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java:85]: From the method name I would read the following information -> we throw an exception if the user is correct?
I think this method needs another name.

**Review by KochTobi — COMMENTED** [review]
Some comments for reference

File: webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java [hunk 11]
```diff
@@ -39,48 +39,53 @@ public void handle(UserRegistrationLayout registrationLayout) {
   }
 
   private void initFields() {
+    userRegistrationLayout.fullName.setPattern(".{1,}");
```
Comment (KochTobi) [inline webapp/src/main/java/life/qbic/views/register/UserRegistrationHandler.java:42]: I do not think this validation is correct. We want a non-empty full name. That requires at least one non-whitespace character.
```suggestion
    userRegistrationLayout.fullName.setPattern("\\S.*");
```
With the requirement: Your name must start with a non-whitespace character.
If we want to allow whitespace in front of the name (and strip it later .. although I do not know why we would do that) the expression would be
```suggestion
    userRegistrationLayout.fullName.setPattern("\\s*\\S.*");
```

**Review by KochTobi — CHANGES_REQUESTED** [review]

File: domain/pom.xml [hunk 12]
```diff
@@ -96,8 +96,9 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <source>16</source>
-                    <target>16</target>
+                    <source>17</source>
+                    <target>17</target>
+                  <compilerArgs>--enable-preview</compilerArgs>
```
Comment (sven1103) [inline domain/pom.xml:101]: Preview features in production code? For what feature do we require Java's preview?

Comment (Steffengreiner) [reply domain/pom.xml:109]: I think this was added by me to try out the new [switch](https://blog.adamgamboa.dev/switch-expression-on-java-17/ ) feature of java 17

File: domain/pom.xml [hunk 13]
```diff
@@ -104,8 +104,9 @@
...
```
Comment (KochTobi) [inline domain/pom.xml:109]:
```suggestion
```

File: domain/src/main/java/life/qbic/domain/user/UserDomainService.java [hunk 14]
```diff
@@ -25,22 +28,24 @@ public UserDomainService(UserRepository userRepository) {
...
-   * <p>Note: this will create a new domain event of type {@link UserRegistered}
+   * <p>Note: this will create a new domain event of type {@link UserRegistered}.
+   * This method is idempotent, if a user already exists with the given emailAddress, no new user
+   * will be created.
    *
    * @param fullName    the full name of the user
-   * @param email       a valid email address
-   * @param rawPassword the raw password desired by the user
+   * @param emailAddress       a valid email address
+   * @param password the encrypted password desired by the user
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/user/UserDomainService.java:37]:
```suggestion
   * @param password the password desired by the user
```

**Review by KochTobi — DISMISSED** [review]
Looks good. However, I noticed that the error shown in case some inputs are not provided correctly does not fit the use case. The invalid credentials error is specific to the login. In the registration case, I do not provide any credentials but only some (personal) information. I would argue it suffices to show the errors on the fields and generalise the text to something like `Please check your input and try again`.
image
